### ISSUE TITLE
Set with user type requires defining two methods

### ISSUE THREAD

**Issue by IainNZ** [JuliaLang/julia#14719]
Its unclear what needs to be defined to make a `Set` work with custom types, but based on this line in the [manual](http://docs.julialang.org/en/latest/stdlib/collections/#iterable-collections):
> `Set`s check whether the item `isequal()`

I figured it was `isequal`. However:

```julia
type Foo; x::Int; end
Base.isequal(a::Foo,b::Foo) = (a.x == b.x)
s = Set(); push!(s, Foo(1)); push!(s, Foo(1))
@show s  # s = Set(Any[Foo(1),Foo(1)])
```

and 

```julia
type Foo; x::Int; end
importall Base.Operators
==(a::Foo,b::Foo) = (a.x == b.x)
s = Set(); push!(s, Foo(1)); push!(s, Foo(1))
@show s  # s = Set(Any[Foo(1),Foo(1)])
```

Apparently the key is you need to define `hash` as well (and `hash` is not sufficient by itself either), e.g.

```julia
type Foo; x::Int; end
Base.isequal(a::Foo,b::Foo) = (a.x == b.x)
Base.hash(f::Foo, h::UInt64) = Base.hash(f.x, h)
s = Set(); push!(s, Foo(1)); push!(s, Foo(1))
@show s  # s = Set(Any[Foo(1)])
```

Having to define `hash` seems cumbersome and unintuitive. In my actual use case the field `x::Vector{Float64}` and I want to be able to define an approximate equality - a notion that doesn't work great with `hash`. Is it possible to just use `isequal`?

**Comment by mbauman** [comment]
#12198

**Comment by IainNZ** [comment]
Ah yes, now I remember that!
I guess https://github.com/JuliaLang/julia/pull/11794#issuecomment-113861001 is relevant too

**Comment by IainNZ** [comment]
I guess maybe a more specific example might be helpful. Essentially what I have is

```julia
type Foo
  x::Float64
end
Base.isequal(a::Foo, b::Foo) = (abs(a.x - b.x) <= 1e-4)
```

and its not even clear to me how one is supposed to write a `hash` function for this. I guess I'm just abusing `Set` too much, as you could have `A==B` and `B==C` but not `A==C` (not in my application, but, in general). Might need to make my own container type. Should I close this in favor of #12198 then?

**Comment by vtjnash** [comment]
It sounds like you are trying to implement interval arithmetic without dealing with the actual implications of that. In particular, your equality function would violate the basic transitivity axiom, so you might need to re-derive all of logic/arithmetic/set-theory/etc. without this assumption.

**Comment by IainNZ** [comment]
Trust me when I tell you that I'm not trying to reimplement interval arithmetic and set theory. Seems like this isn't going to be productive. Closed in favor of #12198.
